### DISCUSSION TITLE
Can Uno.Wasm.Bootstrap be used to compile c# to .wasm?

### DISCUSSION THREAD

**Discussion by nextfool** [unoplatform/Uno.Wasm.Bootstrap#431] [Q&A] [answered]
Hi community,
Sorry I know that asking a questoin by opening an issue may not be a proper way, but I really need help/advice here as I found this project is  by far the most promsing solution to my problem.

So my problem is simple(or complicated), I need to compile my algorithms/libraries written in C#  to .wasm so that they can be used in browser environments(web apps).

So far I have looked in to the blazor webassembly as well as uno platform, but both seems to only emmit a large number of files (dll.bc, dll.clr etc.) other than .wasm.  Can anyone please give me any advice  to put me to the right direction?

Thank you very much!

**Comment by jeromelaban** [comment, marked as answer]
Thanks for asking!

The current tooling for .NET (which the uno bootstrapper uses) does not build a single `.wasm` module for multiple reasons, where the basic one being that emscripten needs additional files to work properly.

Upcoming WebAssembly specs will help removing this dependency (Exceptions, InterfaceTypes) but for now it's not possible.

Added to this, we're currently not packaging code in such a way that it can easily be used as a library, unless the startup code is executed properly. After that, invoking C# code still needs to go through the `C` API of mono, which is not particularly user friendly.

**Reply by nextfool** [reply]
thank you very much for the clarifying!

**Reply by SongOfYouth** [reply]
> Thanks for asking!
> 
> The current tooling for .NET (which the uno bootstrapper uses) does not build a single `.wasm` module for multiple reasons, where the basic one being that emscripten needs additional files to work properly.
> 
> Upcoming WebAssembly specs will help removing this dependency (Exceptions, InterfaceTypes) but for now it's not possible.
> 
> Added to this, we're currently not packaging code in such a way that it can easily be used as a library, unless the startup code is executed properly. After that, invoking C# code still needs to go through the `C` API of mono, which is not particularly user friendly.

Is it possible now?

**Reply by jeromelaban** [reply]
At present time, the only new way to do this is to use Microsoft's experimental Wasi support, but this still is not supported as part of browser wasm implementation, Javascript and emscripten is still needed at this time.

**Reply by SongOfYouth** [reply]
ok.thanks for the reply.

**Reply by BradChase2011** [reply]
> Added to this, we're currently not packaging code in such a way that it can easily be used as a library, unless the startup code is executed properly. After that, invoking C# code still needs to go through the `C` API of mono, which is not particularly user friendly.

I am currently working on trying to get our JS/C/CPP code moved over into our engine project and out of entry point executable project.  Is this saying that there currently isnt a way to have a separate library or project using uno.wasm.bootstap that can be referenced from the startup project?  We currently are getting this compile time error "... is missing an entry point. Add <OutputType>Exe</OutputType> in the project file and a static main.".  If that is what you are saying, can you elaborate on the "executed properly" part so we might determine what to do here?

Hopefully I am not hijacking the thread here, if so please tell me and I can start a new discussion on incorporating uno.wasm.bootstrap into a project with OutputType of Library.

**Reply by jeromelaban** [reply]
@BradChase2011 At this time, no. wasm libraries are not supported in the way uno and .NET are setup. The problem lies in some of the globals that are set that may be conflicting with the existing environment. You could try with the embedded mode, but it's going to be behaving like a singleton.
